libkbfs/coalescing_context.go

```diff
@@ -0,0 +1,110 @@
+package libkbfs
+
+import (
+	"golang.org/x/net/context"
+	"reflect"
+)
+
+// CoalescingContext allows many contexts to be treated as one.  It waits on
+// all its contexts' Context.Done() channels, and when all of them have
+// returned, this CoalescingContext is canceled. At any point, a context can be
+// added to the list, and will subsequently also be part of the wait condition.
+// TODO: add timeout channel in case there is a goroutine leak
+type CoalescingContext struct {
+	context.Context
+	closeCh  chan struct{}
+	doneCh   chan struct{}
+	mutateCh chan context.Context
+	selects  []reflect.SelectCase
+}
+
+func (ctx *CoalescingContext) loop() {
+	for {
+		chosen, val, _ := reflect.Select(ctx.selects)
+		switch chosen {
+		case 0:
```

Comment:
Might be nice to have named constants for these special indices (and for the "empty" length of the slice), and use them both here and when initializing the slice.  Up to you though.